File: samples/orders-sample/src/Orders.Domain/DomainRuleViolation.cs
namespace Orders.Domain;

/// <summary>Violação de regra de negócio. Carrega o ID da regra (docs/regras) para rastreabilidade.</summary>
public sealed class DomainRuleViolationException : Exception
{
    public string RuleId { get; } = string.Empty;

    public DomainRuleViolationException()
    {
    }

    public DomainRuleViolationException(string message) : base(message)
    {
    }

    public DomainRuleViolationException(string message, Exception innerException) : base(message, innerException)
    {
    }

    public DomainRuleViolationException(string ruleId, string message) : base($"{ruleId}: {message}")
    {
        RuleId = ruleId;
    }
}

File: samples/enrichment-lab/src/domain/errors/domain-rule-violation.ts
import { DomainError } from './domain-error';

/**
 * Recusa por regra de negócio. `ruleId` é o ID do bloco em
 * docs/regras/enriquecimento.md — o mesmo que está
 * no nome do teste. Só o domínio lança (ADR-0004); o worker captura, loga `{ ruleId,
 * motivo }` e faz ack:
 * descarte por regra não é retry.
 */
export class DomainRuleViolation extends DomainError {
  constructor(
    readonly ruleId: string,
    readonly motivo: string,
    message = `${ruleId}: ${motivo}`,
  ) {
    super(message, 422);
  }
}
